File: app/services/configuration_service.py
from typing import List, Optional
from datetime import datetime
from fastapi import HTTPException, status
from app.models.configuration import (
    Configuration,
    ConfigurationCreate,
    ConfigurationUpdate,
    ConfigurationQuery,
    BackupResponse,
)
from app.repositories.configuration_repository import ConfigurationRepository
from app.core.logger import logger


class ConfigurationService:
    def __init__(self, repository: ConfigurationRepository):
        self.repository = repository

    async def create(self, dto: ConfigurationCreate, user_id: str) -> Configuration:
        """Create new configuration"""
        try:
            # Check if key already exists
            existing = await self.repository.find_by_key(dto.key, dto.tenant_id)
            if existing:
                raise HTTPException(
                    status_code=status.HTTP_409_CONFLICT,
                    detail=f"Configuration with key '{dto.key}' already exists",
                )

            config = Configuration(
                **dto.model_dump(by_alias=True), created_by=user_id, updated_by=user_id
            )
            result = await self.repository.create(config)

            logger.info(f"Configuration created: {result.id} by user {user_id}")
            return result
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error in create configuration: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)
            )

    async def get_by_id(
        self, config_id: str, tenant_id: str, include_inherited: bool = False
    ) -> Configuration:
        """Get configuration by ID"""
        try:
            # Fetch from database
            config = await self.repository.find_by_id(config_id, tenant_id)
            if not config:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND, detail="Configuration not found"
                )

            # Apply inheritance if needed
            if include_inherited and config.parent_config_id:
                config = await self._apply_inheritance(config)

            return config
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error in get_by_id: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)
            )

    async def get_by_key(
        self, key: str, tenant_id: str, include_inherited: bool = False
    ) -> Configuration:
        """Get configuration by key"""
        try:
            config = await self.repository.find_by_key(key, tenant_id)
            if not config:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND,
                    detail=f"Configuration with key '{key}' not found",
                )

            if include_inherited and config.parent_config_id:
                config = await self._apply_inheritance(config)

            return config
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error in get_by_key: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)
            )

    async def list(self, query: ConfigurationQuery) -> List[Configuration]:
        """List configurations"""
        try:
            configs = await self.repository.find_all(query)

            if query.include_inherited:
                processed_configs = []
                for config in configs:
                    if config.parent_config_id:
                        config = await self._apply_inheritance(config)
                    processed_configs.append(config)
                return processed_configs

            return configs
        except Exception as e:
            logger.error(f"Error in list: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)
            )

    async def update(
        self, config_id: str, tenant_id: str, dto: ConfigurationUpdate, user_id: str
    ) -> Configuration:
        """Update configuration"""
        try:
            existing = await self.repository.find_by_id(config_id, tenant_id)
            if not existing:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND, detail="Configuration not found"
                )

            updates = dto.model_dump(by_alias=True, exclude_unset=True)
            updates["updated_by"] = user_id

            updated = await self.repository.update(config_id, tenant_id, updates)

            logger.info(f"Configuration updated: {config_id} by user {user_id}")
            return updated
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error in update: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)
            )

    async def delete(self, config_id: str, tenant_id: str) -> None:
        """Delete configuration"""
        try:
            existing = await self.repository.find_by_id(config_id, tenant_id)
            if not existing:
                raise HTTPException(
                    status_code=status.HTTP_404_NOT_FOUND, detail="Configuration not found"
                )

            await self.repository.delete(config_id, tenant_id)

            logger.info(f"Configuration deleted: {config_id}")
        except HTTPException:
            raise
        except Exception as e:
            logger.error(f"Error in delete: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)
            )

    async def backup(
        self, tenant_id: str, user_id: str, description: Optional[str] = None
    ) -> BackupResponse:
        """Create backup of configurations"""
        try:
            configs = await self.repository.find_all(ConfigurationQuery(tenant_id=tenant_id))

            backup_id = f"backup-{tenant_id}-{int(datetime.utcnow().timestamp())}"
            backup_data = {
                "id": backup_id,
                "tenant_id": tenant_id,
                "backup_date": datetime.utcnow().isoformat(),
                "configurations": [c.model_dump(by_alias=True, mode="json") for c in configs],
                "created_by": user_id,
                "description": description,
            }

            # Note: Without Redis, backup functionality is limited
            # Consider implementing persistent backup storage if needed
            logger.warning("Backup feature requires persistent storage implementation")
            logger.info(f"Backup created for tenant {tenant_id}: {backup_id}")
            return BackupResponse(backup_id=backup_id)
        except Exception as e:
            logger.error(f"Error in backup: {e}")
            raise HTTPException(
                status_code=status.HTTP_500_INTERNAL_SERVER_ERROR, detail=str(e)
            )

    async def restore(self, backup_id: str, tenant_id: str, user_id: str) -> None:
        """Restore configurations from backup"""
        # Note: Without Redis, restore functionality is limited
        # Consider implementing persistent backup storage if needed
        logger.warning("Restore feature requires persistent storage implementation")
        raise HTTPException(
            status_code=status.HTTP_501_NOT_IMPLEMENTED,
            detail="Restore feature requires persistent storage. Please implement backup storage."
        )

    async def _apply_inheritance(self, config: Configuration) -> Configuration:
        """Apply configuration inheritance"""
        if not config.parent_config_id:
            return config

        try:
            parent = await self.repository.find_by_id(config.parent_config_id, config.tenant_id)
            if not parent:
                logger.warning(f"Parent configuration not found: {config.parent_config_id}")
                return config

            # Merge parent values with current config (current config takes precedence)
            inherited_dict = parent.model_dump(by_alias=True)
            config_dict = config.model_dump(by_alias=True)
            
            # Override with child values
            inherited_dict.update({k: v for k, v in config_dict.items() if v is not None})
            inherited_dict["id"] = config.id
            inherited_dict["key"] = config.key
            inherited_dict["value"] = config.value

            return Configuration(**inherited_dict)
        except Exception as e:
            logger.error(f"Error applying inheritance: {e}")
            return config

File: src/service-setting-service/src/services/configurationService.ts
import { ConfigurationRepository } from '../repositories/configurationRepository';
import { CacheService } from './cacheService';
import {
  Configuration,
  CreateConfigurationDto,
  UpdateConfigurationDto,
  ConfigurationQuery,
} from '../types';
import logger from '../utils/logger';
import { AppError } from '../middleware/errorHandler';

export class ConfigurationService {
  private repository: ConfigurationRepository;
  private cache: CacheService;

  constructor(repository: ConfigurationRepository, cache: CacheService) {
    this.repository = repository;
    this.cache = cache;
  }

  private getCacheKey(id: string, tenantId: string): string {
    return `config:${tenantId}:${id}`;
  }

  private getTenantCachePattern(tenantId: string): string {
    return `config:${tenantId}:*`;
  }

  async create(dto: CreateConfigurationDto, userId: string): Promise<Configuration> {
    try {
      // Check if key already exists
      const existing = await this.repository.findByKey(dto.key, dto.tenantId);
      if (existing) {
        throw new AppError(409, `Configuration with key '${dto.key}' already exists`);
      }

      const config = await this.repository.create({
        ...dto,
        isEncrypted: dto.isEncrypted || false,
        createdBy: userId,
        updatedBy: userId,
      });

      // Cache the new configuration
      await this.cache.set(this.getCacheKey(config.id, config.tenantId), config);

      logger.info(`Configuration created: ${config.id} by user ${userId}`);
      return config;
    } catch (error) {
      logger.error('Error in create configuration:', error);
      throw error;
    }
  }

  async getById(id: string, tenantId: string, includeInherited = false): Promise<Configuration> {
    try {
      // Try cache first
      const cacheKey = this.getCacheKey(id, tenantId);
      const cached = await this.cache.get<Configuration>(cacheKey);
      if (cached) {
        logger.debug(`Configuration ${id} retrieved from cache`);
        return cached;
      }

      // Fetch from database
      const config = await this.repository.findById(id, tenantId);
      if (!config) {
        throw new AppError(404, 'Configuration not found');
      }

      // Apply inheritance if needed
      let finalConfig = config;
      if (includeInherited && config.parentConfigId) {
        finalConfig = await this.applyInheritance(config);
      }

      // Cache the result
      await this.cache.set(cacheKey, finalConfig);

      return finalConfig;
    } catch (error) {
      logger.error('Error in getById:', error);
      throw error;
    }
  }

  async getByKey(key: string, tenantId: string, includeInherited = false): Promise<Configuration> {
    try {
      const config = await this.repository.findByKey(key, tenantId);
      if (!config) {
        throw new AppError(404, `Configuration with key '${key}' not found`);
      }

      let finalConfig = config;
      if (includeInherited && config.parentConfigId) {
        finalConfig = await this.applyInheritance(config);
      }

      return finalConfig;
    } catch (error) {
      logger.error('Error in getByKey:', error);
      throw error;
    }
  }

  async list(query: ConfigurationQuery): Promise<Configuration[]> {
    try {
      const configs = await this.repository.findAll(query);

      if (query.includeInherited) {
        const processedConfigs = await Promise.all(
          configs.map((config) =>
            config.parentConfigId ? this.applyInheritance(config) : config
          )
        );
        return processedConfigs;
      }

      return configs;
    } catch (error) {
      logger.error('Error in list:', error);
      throw error;
    }
  }

  async update(
    id: string,
    tenantId: string,
    dto: UpdateConfigurationDto,
    userId: string
  ): Promise<Configuration> {
    try {
      const existing = await this.repository.findById(id, tenantId);
      if (!existing) {
        throw new AppError(404, 'Configuration not found');
      }

      const updated = await this.repository.update(id, tenantId, {
        ...dto,
        updatedBy: userId,
      });

      // Invalidate cache
      await this.cache.delete(this.getCacheKey(id, tenantId));

      logger.info(`Configuration updated: ${id} by user ${userId}`);
      return updated;
    } catch (error) {
      logger.error('Error in update:', error);
      throw error;
    }
  }

  async delete(id: string, tenantId: string): Promise<void> {
    try {
      const existing = await this.repository.findById(id, tenantId);
      if (!existing) {
        throw new AppError(404, 'Configuration not found');
      }

      await this.repository.delete(id, tenantId);

      // Invalidate cache
      await this.cache.delete(this.getCacheKey(id, tenantId));

      logger.info(`Configuration deleted: ${id}`);
    } catch (error) {
      logger.error('Error in delete:', error);
      throw error;
    }
  }

  async backup(tenantId: string, userId: string, description?: string): Promise<{ backupId: string }> {
    try {
      const configs = await this.repository.findAll({ tenantId });

      const backupId = `backup-${tenantId}-${Date.now()}`;
      const backup = {
        id: backupId,
        tenantId,
        backupDate: new Date(),
        configurations: configs,
        createdBy: userId,
        description,
      };

      // Store backup in cache (could also store in a separate container)
      await this.cache.set(`backup:${backupId}`, backup, 86400 * 7); // 7 days TTL

      logger.info(`Backup created for tenant ${tenantId}: ${backupId}`);
      return { backupId };
    } catch (error) {
      logger.error('Error in backup:', error);
      throw error;
    }
  }

  async restore(backupId: string, tenantId: string, userId: string): Promise<void> {
    try {
      const backup = await this.cache.get<{
        configurations: Configuration[];
        tenantId: string;
      }>(`backup:${backupId}`);

      if (!backup) {
        throw new AppError(404, 'Backup not found or expired');
      }

      if (backup.tenantId !== tenantId) {
        throw new AppError(403, 'Cannot restore backup from different tenant');
      }

      // Restore configurations
      for (const config of backup.configurations) {
        const existing = await this.repository.findById(config.id, tenantId);
        if (existing) {
          await this.repository.update(config.id, tenantId, {
            ...config,
            updatedBy: userId,
          });
        } else {
          await this.repository.create({
            ...config,
            createdBy: userId,
            updatedBy: userId,
          });
        }
      }

      // Invalidate all tenant cache
      await this.cache.deletePattern(this.getTenantCachePattern(tenantId));

      logger.info(`Backup restored for tenant ${tenantId}: ${backupId}`);
    } catch (error) {
      logger.error('Error in restore:', error);
      throw error;
    }
  }

  private async applyInheritance(config: Configuration): Promise<Configuration> {
    if (!config.parentConfigId) {
      return config;
    }

    try {
      const parent = await this.repository.findById(config.parentConfigId, config.tenantId);
      if (!parent) {
        logger.warn(`Parent configuration not found: ${config.parentConfigId}`);
        return config;
      }

      // Merge parent values with current config (current config takes precedence)
      const inherited: Configuration = {
        ...parent,
        ...config,
        id: config.id,
        key: config.key,
        value: config.value !== undefined ? config.value : parent.value,
      };

      return inherited;
    } catch (error) {
      logger.error('Error applying inheritance:', error);
      return config;
    }
  }
}
